Lib/test/test_unittest/test_runner.py

```diff
@@ -1452,6 +1452,206 @@ def test_1(self):
         run(Foo('test_1'), expect_durations=False)
 
 
+class SubTestMockContextManager:
+    def __init__(self, events_list, name="cm"):
+        self.events = events_list
+        self.name = name
+        self.entered_value = f"{name}_entered_value"
+
+    def __enter__(self):
+        self.events.append(f"{self.name}_enter")
+        return self.entered_value
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.events.append((f"{self.name}_exit", exc_type.__name__ if exc_type else None))
+        return False
+
+
+class SubTestFailingEnterCM:
+    def __init__(self, events_list, name="cm_enter_fail"):
+        self.events = events_list
+        self.name = name
+
+    def __enter__(self):
+        self.events.append(f"{self.name}_enter_attempt")
+        raise RuntimeError("Simulated __enter__ failure")
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.events.append((f"{self.name}_exit_UNEXPECTED", exc_type.__name__ if exc_type else None))
+        return False
+
+
+class TestSubTestCleanups(unittest.TestCase):
+    def setUp(self):
+        self.events = []
+        TestSubTestCleanups._active_events_list = self.events
+        TestSubTestCleanups._active_instance = self
+
+    def tearDown(self):
+        if hasattr(TestSubTestCleanups, '_active_events_list'):
+            del TestSubTestCleanups._active_events_list
+        if hasattr(TestSubTestCleanups, '_active_instance'):
+            del TestSubTestCleanups._active_instance
+
+    @classmethod
+    def _get_events(cls):
+        return cls._active_events_list
+
+    def _record_event(self, event_name, *args, **kwargs):
+        TestSubTestCleanups._get_events().append(event_name)
+
+    def test_addCleanup_operation_and_LIFO_order(self):
+        class MyTests(unittest.TestCase):
+            def runTest(inner_self):
+                events = TestSubTestCleanups._get_events()
+                recorder = TestSubTestCleanups._active_instance._record_event
+                with inner_self.subTest() as sub:
+                    events.append("subtest_body_start")
+                    sub.addCleanup(recorder, "cleanup_2_args", "arg")
+                    sub.addCleanup(recorder, "cleanup_1")
+                    events.append("subtest_body_end")
+
+        MyTests().run()
+
+        expected_events = [
+            "subtest_body_start",
+            "subtest_body_end",
+            "cleanup_1",
+            "cleanup_2_args",
+        ]
+        self.assertEqual(self.events, expected_events)
```

Comment:
There should be no need for `_active_events_list` and `_get_events` and `_record_event`.
```suggestion
    def test_addCleanup_operation_and_LIFO_order(self):
        events = []
        class MyTests(unittest.TestCase):
            def runTest(inner_self):
                def record(event_name, *args, **kwargs):
                    events.append((event_name, args, kwargs))
                with inner_self.subTest() as sub:
                    events.append("subtest_body_start")
                    sub.addCleanup(record, "cleanup_2_args", "pos", kw="kwd")
                    sub.addCleanup(record, "cleanup_1")
                    events.append("subtest_body_end")

        MyTests().run()

        expected_events = [
            "subtest_body_start",
            "subtest_body_end",
            ("cleanup_1", (), {}),
            ("cleanup_2_args", ("pos", ), {"kw": "kwd"}),
        ]
        self.assertEqual(events, expected_events)
```